src/core/experimental/array.d

```diff
@@ -0,0 +1,1480 @@
+    immutable(rcarray!T) idup(this Q)()
+    {
+        static if (is(Q == immutable))
+        {
+            return this;
+        }
+        else
+        {
+            return immutable rcarray!T(this);
+        }
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        {
+            auto a = rcarray!int(1, 2, 3);
+            auto a2 = a.idup();
+            static assert (is(typeof(a2) == immutable));
+        }
+
+        {
+            auto a = const rcarray!int(1, 2, 3);
+            auto a2 = a.idup();
+            static assert (is(typeof(a2) == immutable));
+        }
+
+        {
+            auto a = immutable rcarray!int(1, 2, 3);
+            auto a2 = a.idup();
+            static assert (is(typeof(a2) == immutable));
+        }
+    }
+
+    /**
+     * Perform a copy of the array. This will create a new array that will copy
+     * the elements of the current array. This will `NOT` call `dup` on the
+     * elements of the array, regardless if `T` defines it or not.
+     *
+     * Returns:
+     *      a new mutable array.
+     *
+     * Complexity: $(BIGOH n).
+     */
+    rcarray!T dup(this Q)()
+    {
+        return rcarray!T(payload);
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        enum stuff = [1, 2, 3];
+        auto a = immutable rcarray!int(stuff);
+        auto aDup = a.dup;
+        assert(aDup == stuff);
+        aDup[0] = 0;
+        assert(aDup[0] == 0);
+        assert(a[0] == 1);
+    }
+
+    /**
+     * Return a slice to the current array. This is equivalent to performing
+     * a shallow copy of the array.
+     *
+     * Returns:
+     *      an array that references the current array.
+     *
+     * Complexity: $(BIGOH 1)
+     */
+    auto opSlice(this Qualified)()
+    {
+        return this;
+    }
+
+    /**
+     * Return a slice to the current array that is bounded by `start` and `end`.
+     * `start` must be less than or equal to `end` and `end` must be less than
+     * or equal to `length`.
+     *
+     * Returns:
+     *      an array that references the current array.
+     *
+     * Params:
+     *      start = a positive integer
+     *      end = a positive integer
+     *
+     * Complexity: $(BIGOH 1)
+     */
+    auto opSlice(this Qualified)(size_t start, size_t end)
+    {
+        return typeof(this)(rc, support, payload[start .. end]);
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto stuff = [1, 2, 3];
+        auto a = immutable rcarray!int(stuff);
+        assert(a[] == stuff);
+        assert(a[1 .. $] == stuff[1 .. $]);
+    }
+
+    /**
+     * Provide access to the element at `idx` in the array.
+     * `idx` must be less than `length`.
+     *
+     * Returns:
+     *      a reference to the element found at `idx`.
+     *
+     * Params:
+     *      idx = a positive integer
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    ref auto opIndex(this _)(size_t idx)
```

Comment:
What's `this _` for?

Reply:
Attribute inference. I could make it `inout` instead?

Reply:
Yes, it `inout` should be used exactly in this case: when the mutability qualifier of the return type needs to be same as the one of `this`.

Reply:
Otherwise, to make a dummy template, it’s enough to add an extra set of empty parentheses.